File: src/core/ports/inbound/viewer.ts
import type { InputEvent } from '../../domain/input/events.js';
import type { ScreenFrame, ScreenSize } from '../../domain/screen/frame.js';
import type { ConnectOptions, SessionInfo, SessionStats } from '../../domain/session/config.js';

export type SessionEvent =
  | { kind: 'frame'; frame: ScreenFrame }
  | { kind: 'screen'; screen: ScreenSize };

export interface SessionPort {
  info(): SessionInfo;
  events(): AsyncIterable<SessionEvent>;
  sendInput(event: InputEvent): void;
  stats(): SessionStats;
  close(reason?: string): Promise<void>;
}

export interface ViewerPort {
  connect(options: ConnectOptions): Promise<SessionPort>;
}

File: internal/port/viewer.go
package port

import (
	"context"

	"ghostwire/internal/domain/input"
	"ghostwire/internal/domain/screen"
)

type SessionEvent struct {
	Frame *screen.ScreenFrame
	Screen *screen.ScreenSize
}

type SessionInfo struct {
	Screen        screen.ScreenSize
	Fps           int
	ServerVersion string
}

type SessionStats struct {
	FramesReceived int
	FramesDropped  int
	BytesReceived  int
	RttMs          int64
	Fps            int
}

type SessionPort interface {
	Info() SessionInfo
	Events() <-chan SessionEvent
	SendInput(event input.InputEvent)
	Stats() SessionStats
	Close(reason string) error
}

type ViewerPort interface {
	Connect(ctx context.Context, opts ConnectOptions) (SessionPort, error)
}
